File: patterns/vision-and-images/gemini/javascript/image-analysis.ts
import 'dotenv/config';
import { GoogleGenerativeAI } from '@google/generative-ai';

/**
 * Demonstrates sending an image to Gemini 1.5 Flash for analysis.
 * Note: In Node.js, we typically read files or use fetch to get buffers.
 */
async function main() {
    const apiKey = process.env.GEMINI_API_KEY;
    if (!apiKey) {
        console.error('Please set GEMINI_API_KEY.');
        return;
    }

    const genAI = new GoogleGenerativeAI(apiKey);
    const model = genAI.getGenerativeModel({ model: 'gemini-1.5-flash' });

    const imageUrl =
        'https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg';

    console.log(`Fetching image from: ${imageUrl}...`);

    try {
        // 1. Fetch Image
        const response = await fetch(imageUrl);
        const arrayBuffer = await response.arrayBuffer();
        const buffer = Buffer.from(arrayBuffer);

        // 2. Prepare Part
        const imagePart = {
            inlineData: {
                data: buffer.toString('base64'),
                mimeType: 'image/jpeg',
            },
        };

        // 3. Generate Content
        console.log("Analyzing image with Gemini...");
        const result = await model.generateContent([
            "Describe this image specifically focusing on the colors.",
            imagePart
        ]);
        const text = result.response.text();

        console.log('\nDescription:');
        console.log(text);
    } catch (error) {
        console.error('Error:', error);
    }
}

main();

File: patterns/vision-and-images/gemini/python/image_analysis.py
import os
import requests
import google.generativeai as genai
from PIL import Image
from io import BytesIO
from dotenv import load_dotenv

load_dotenv()

def main():
    """
    Demonstrates sending an image to Gemini 1.5 Flash for analysis.
    Downloads an image from a URL and passes it to the model.
    """
    api_key = os.environ.get("GEMINI_API_KEY")
    if not api_key:
        print("Please set GEMINI_API_KEY.")
        return

    genai.configure(api_key=api_key)

    # 1. Download image
    image_url = "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg"
    print(f"Downloading image from: {image_url}...")
    
    response = requests.get(image_url)
    img = Image.open(BytesIO(response.content))

    # 2. Initialize Model
    model = genai.GenerativeModel('gemini-1.5-flash')

    print("Analyzing image with Gemini...")

    # 3. Generate Content
    response = model.generate_content(["Describe this image specifically focusing on the colors.", img])

    print("\nDescription:")
    print(response.text)

if __name__ == "__main__":
    main()
